Add regex support in Auto Color for track, marker, region name filters

daleonov opened the pull request:
Exact matching didn't really work for track names for me. So I added regex support to define a pattern, and keywords like "gtr", "bass", "synth" won't conflict with "lead", "rhythm", "double" etc and each combo would be able to have its own color.
![sws regex](https://user-images.githubusercontent.com/5140018/72608192-3c89ca80-3933-11ea-843a-beea3672b4cb.png)

daleonov commented:
I can also look into #1257 since it's about similar feature.

nofishonfriday commented:
I was unaware of this either until [shortly ago](https://github.com/reaper-oss/sws/pull/1230#issuecomment-562285212) but since SWS targets macOS 10.5. the C++11 regex features are not available for the Mac version, so we can't use them unfortunately (afaik).
https://travis-ci.com/reaper-oss/sws/jobs/276788893#L546

cfillion commented:
In addition to the above, the `std::regex` constructor throws an `std::invalid_regex` exception when the input is invalid. Since it comes from users, the possibility of an invalid regex should be handled without crashing.

Currently it bricks REAPER by making it crash on startup until the invalid regex is manually removed from the .ini file.

I find the `$` prefix somewhat confusing too, as it is usually the end of string/line anchor. Perhaps surrounding the regexes with slashes would make it clearer? It is a widely used delimiter for regexes.

Additional concerns I'm not sure about (needs testing):
- Cost of compiling regexes over and over with lots of tracks
- Possible hangs with input accidentally resulting in overly complex regexes?
- What are the advantages of using the POSIX extended syntax vs the default ECMAScript?

As for implementing #1257 (and the related #1119), that can be done by reusing ReaPack's filter implementation (but with its use of the C++17 stdlib removed) as discussed [here](https://github.com/reaper-oss/sws/issues/1119#issuecomment-493083501).

daleonov commented:
> the C++11 regex features are not available for the Mac version, so we can't use them unfortunately (afaik).

Can't use them at all or still can use as a platform-specific feature? The code is easy to isolate so I can do something like
```
#ifndef GRANDPA_TECH
do_regex();
#endif 
```
...if such thing is acceptable here, of course.
Other way is using third-party library, but I don't think it's welcome here unless extremely necessary?

> the std::regex constructor throws an std::regex_error exception when the input is invalid
Fair enough. No elegant way to address it other than catching it? Good news is `std::regex_error` is easy to isolate.

> Perhaps surrounding the regexes with slashes would make it clearer?
Good idea!

> Cost of compiling regexes over and over with lots of tracks (vs storing a compiled `std::regex` and reusing it)
Fair enough. It's a lower priority concern than the rest of those. But if I go with regex exception handling (as I discussed above) then I'll have one more reason to move it outside the loop.

> Possible hangs with user input accidentally resulting in overly complex regexes?
Possible... But it's their DAW, right? I think it's up to them to decide how complex is too complex :)

> What are the advantages of using the POSIX extended syntax vs the default ECMAScript?
There was something I didn't like, so I always used the extended one. Don't remember exactly, I'll check later. As long as it can do what's on screenshot in first post, I'm happy.

> Wouldn't the `collate` option introduce behavior differences from an install to the next depending on the current system locale?
And I think it should. What I had in mind is stuff like `\w` including user's alphabet. I haven't tested it yet but I assumed it's what this flag does.

Also, I've hardcoded it to be case insensitive, not quite happy with it. I don't recall anything in actual regex syntax to specify case sensitivity, and I don't see any elegant way to specify it in the sws enviroment. Other than adding another meta language thing to filter string, but that's clunky too.

----

Okay, so let's first decide if it's possible at all to add this feature (since there are some macOS compatibility issues) and if it is, I'll proceed with addressing the points above. Hopefully I won't have to fork it just for that little feature :)

cfillion commented:
> Can't use them at all or still can use as a platform-specific feature?

It wouldn't be the first platform-specific feature in SWS (there are some Windows-only ones). It would be a shame to have this everywhere except on macOS though...

> Other way is using third-party library, but I don't think it's welcome here unless extremely necessary?

Pretty much. Regexes in auto color would be a pretty niche feature so I don't think it's enough to warrant bundling an entire regex library. From what I could find (ignoring compatibility concerns regarding the regex flavor), Apple's Foundation framework added regexes in v10.7 and libpcre is bundled with the OS since 10.6. 10.5 doesn't get any.

> No elegant way to address it other than catching it?

That's correct.

> Possible... But it's their DAW, right?

If they can accidentally brick their DAW (eg. freeze at startup), it might become an issue. (I haven't looked into it so it might be very hard/impossible to so. In which case this is a non-issue.)

> And I think it should. What I had in mind is stuff like \w including user's alphabet.

That would cause issues when using an existing config one a computer that happens to have a different locale than the one it was tested on (eg. cool Auto Color regex tips shared on the forum mysteriously doesn't work, portable install on a thumb drive appears broken at the studio's PC...).

> I don't recall anything in actual regex syntax to specify case sensitivity, and I don't see any elegant way to specify it in the sws enviroment.

How about `/regex string/i` for case-insensitive? Since it's a common syntax used in Perl, JavaScript, Ruby, PHP...

---

Alternatively, instead of regular expressions, perhaps REAPER's Action List syntax would be helpful in Auto Color? It has AND/OR/NOT, grouping with parenthesis, and full-word matching.

It's not very useful alternative with patterns such as `V[0-9]` (becoming the very ugly `V0 OR V1 OR V2 OR...`), but things like `(GTR|guitar).*lead` can be written as `( GTR OR guitar ) lead` (except it allows `lead` to appear anywhere including before `GTR`).

daleonov commented:
Using simpler syntax would be great! I agree, regex is a bit too geeky even for reaper users.
But parsing it without regex and not being able to use regex under the hood would be one hell of a chore... Is something similar already implemented somewhere in sws?

cfillion commented:
In that case this PR, along with #1119 (where it was originally suggested), #1257 and #1285 could all be solved by adding ReaPack's [implementation](https://github.com/cfillion/reapack/blob/master/src/filter.cpp) of it into SWS (with minor tweaks to remove uses of `std::unique_ptr` and `boost::algorithm::to_lower`) .

odedd commented:
Any chance of getting this thing to work? This could be a really powerful addition

sychron commented:
Unfortunately, My c++ FU is not good enaugh to resolve the conflicts. 

But I would really like to have this feature, as it's very handy if you have track names that have the description in the front and a labelling in the back, for example: "FX: <rolename> A", "FX: <rolename> B" etc. With Regex, one could assign track icons like FX:A, FX:B, ignoring the role name.
So ... please count this as an upvote for this request :)
